internal/app/orchestrator/downstream.go

```diff
@@ -14,82 +14,73 @@ package orchestrator
 import (
 	"sync"
 
-	"github.com/envoyproxy/xds-relay/internal/app/mapper"
 	"github.com/envoyproxy/xds-relay/internal/app/transport"
 )
 
 // downstreamResponseMap is a map of downstream xDS client requests to response
 // channels.
 type downstreamResponseMap struct {
 	mu      sync.RWMutex
-	watches map[transport.Request]transport.Watch
+	watches map[string][]transport.Watch
 }
 
 func newDownstreamResponseMap() downstreamResponseMap {
 	return downstreamResponseMap{
-		watches: make(map[transport.Request]transport.Watch),
+		watches: make(map[string][]transport.Watch),
 	}
 }
 
 // createWatch initializes a new channel for a request if it doesn't already
 // exist.
-func (d *downstreamResponseMap) createWatch(req transport.Request) transport.Watch {
+func (d *downstreamResponseMap) addWatch(aggregatedKey string, w transport.Watch) {
 	d.mu.Lock()
 	defer d.mu.Unlock()
-	if _, ok := d.watches[req]; !ok {
-		d.watches[req] = req.CreateWatch()
+	if _, ok := d.watches[aggregatedKey]; !ok {
+		d.watches[aggregatedKey] = make([]transport.Watch, 0, 10)
 	}
-	return d.watches[req]
+	d.watches[aggregatedKey] = append(d.watches[aggregatedKey], w)
 }
 
 // get retrieves the channel where responses are set for the specified request.
-func (d *downstreamResponseMap) get(req transport.Request) (transport.Watch, bool) {
+func (d *downstreamResponseMap) getSnapshot(aggregatedKey string) ([]transport.Watch, bool) {
 	d.mu.RLock()
 	defer d.mu.RUnlock()
-	channel, ok := d.watches[req]
-	return channel, ok
+	watches, ok := d.watches[aggregatedKey]
+	if ok {
+		d.watches[aggregatedKey] = make([]transport.Watch, 0, 10)
+	}
+
+	return watches, ok
+}
+
+// get retrieves the channel where responses are set for the specified request.
+func (d *downstreamResponseMap) get(aggregatedKey string) []transport.Watch {
+	d.mu.RLock()
+	defer d.mu.RUnlock()
+	return d.watches[aggregatedKey]
 }
 
 // delete removes the response channel and request entry from the map and
 // closes the corresponding channel.
-func (d *downstreamResponseMap) delete(req transport.Request) transport.Watch {
-	d.mu.Lock()
-	defer d.mu.Unlock()
-	if watch, ok := d.watches[req]; ok {
-		// wait for all writes to the responseChannel to complete before closing.
-		watch.Close()
-		delete(d.watches, req)
-		return watch
-	}
-	return nil
+func (d *downstreamResponseMap) delete(w transport.Watch) {
+	w.Close()
```

Comment:
Do we need to remove it from the slice as well?

Reply:
we dont do it since removal from slice is expensive. The slice is completely rolled over during fanout.